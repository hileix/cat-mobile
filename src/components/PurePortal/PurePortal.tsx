import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

export interface PurePortalProps {
  /**
   * children，即 ReactDOM.createPortal() 的第一个参数
   */
  children?: React.ReactNode;
  /**
   * 作为容器的 css 选择器（或 dom 元素）
   */
  selector: string | HTMLElement;
  /**
   * unmount 时的回调
   */
  onUnmount?: () => void;
}

/**
 * RaectDOM.createPortal() 的组件化形式
 */
class PurePortal extends React.Component<PurePortalProps> {
  static propTypes = {
    children: PropTypes.node,
    selector: PropTypes.oneOfType([PropTypes.string, PropTypes.object,]),
    render: PropTypes.func,
    onMount: PropTypes.func,
    onUnmount: PropTypes.func
  };
  static defaultProps = {
    selector: 'body'
  };

  private container: Element | null;

  componentDidMount = () => {
    this.getContainer();
  };

  componentDidUpdate = (prevProps: PurePortalProps) => {
    if (prevProps.selector !== this.props.selector) {
      this.getContainer();
    }
  };

  componentWillUnmount = () => {
    const { onUnmount } = this.props;
    this.container = null;
    onUnmount && onUnmount();
  };

  getContainer = (): void => {
    const { selector } = this.props;
    let container;

    if (selector instanceof HTMLElement) {
      container = selector;
    } else {
      container = document.querySelector(selector);
    }

    if (!container) {
      container = document.body;
    }
    this.container = container;
    this.forceUpdate();
  };

  render(): React.ReactPortal | null {
    if (!this.container) {
      return null;
    }


    return ReactDOM.createPortal(this.props.children, this.container);
  }
}

export default PurePortal;
