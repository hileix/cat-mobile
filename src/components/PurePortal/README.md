# PurePortal 纯传送门

## Example

```jsx
<PurePortal selector='body'>
  <div>children</div>
</PurePortal>
```

## API

|   属性   |       类型        | 默认值 |                       描述                        |
|:--------:|:-----------------:|:------:|:-------------------------------------------------:|
| children | `React.ReactNode` |   -    | children，即 ReactDOM.createPortal() 的第一个参数 |
| selector |     `string`      | `body` |                 容器的 css 选择器                 |
