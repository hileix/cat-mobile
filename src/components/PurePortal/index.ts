import PurePortal from './PurePortal';
export default PurePortal;
