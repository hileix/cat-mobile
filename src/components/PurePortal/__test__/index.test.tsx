import React from 'react';
import { mount } from 'enzyme';
import sinon from 'sinon';

import PurePortal from '../index';

describe('PurePortal', () => {
  // reset dom
  beforeEach(() => {
    document.body.innerHTML = '';
  });
  afterEach(() => {
    document.body.innerHTML = '';
  });

  test('renders correctly', () => {
    const wrapper = mount(<PurePortal>content</PurePortal>);
    expect(wrapper).toMatchSnapshot();

    let portalContainerDOM = document.createElement('div');
    portalContainerDOM.className = 'my-portal-container';
    document.body.append(portalContainerDOM);

    mount(<PurePortal selector={portalContainerDOM}>content</PurePortal>)
    expect(portalContainerDOM.innerHTML).toBe('content');

  });

  test('onUnmount is called', () => {
    const onUnmount = sinon.spy();
    const wrapper = mount(
      <PurePortal onUnmount={onUnmount}>content</PurePortal>
    );
    expect(onUnmount.callCount).toBe(0);
    wrapper.unmount();
    expect(onUnmount.callCount).toBe(1);
  });

  test('If the selector does not have a dom corresponding to it, the body is used', () => {
    const wrapper = mount(
      <PurePortal selector='does-not-have-a-dom-selector'>
        <span className='my-pure-portal-content'>content</span>
      </PurePortal>
    );
    let contentDOM = document.querySelector('.my-pure-portal-content');
    let parentDOM = (contentDOM as Element).parentNode;
    expect(parentDOM instanceof HTMLBodyElement).toBe(true);
  });

  test('When a selector is changed, the content is rendered to the selector container', () => {
    const wrapper = mount(
      <PurePortal selector='body'>
        <span className='my-pure-portal-content'>content</span>
      </PurePortal>
    );

    let contentDOM = document.querySelector('.my-pure-portal-content');
    let parentDOM = (contentDOM as any).parentNode;
    expect(parentDOM instanceof HTMLBodyElement).toBe(true);

    // create dom
    const divDOM = document.createElement('div');
    divDOM.setAttribute('class', 'my-new-container');
    document.body.append(divDOM);
    // change selector props
    wrapper.setProps({ selector: '.my-new-container' });

    contentDOM = document.querySelector('.my-pure-portal-content');
    parentDOM = (contentDOM as Element).parentNode;

    expect((parentDOM as Element).getAttribute('class')).toBe(
      'my-new-container'
    );
  });
});
