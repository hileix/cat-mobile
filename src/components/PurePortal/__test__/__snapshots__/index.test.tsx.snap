// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PurePortal renders correctly 1`] = `
<PurePortal
  selector="body"
>
  <Portal
    containerInfo={
      <body>
        content
      </body>
    }
  >
    content
  </Portal>
</PurePortal>
`;
